File: client/soap_client.py
import zeep

URL = "http://localhost:8002/?wsdl"  # Corrigido: endpoint sem /soap
client = zeep.Client(URL)

def listar_tarefas():
    try:
        resposta = client.service.listar_tarefas()
        print("SOAP retorno bruto:", resposta)

        tarefas = []
        for tarefa in resposta:
            # Zeep retorna objetos tipo "zeep.objects.Tarefa", que podem ser convertidos em dict assim:
            tarefa_dict = {
                "id": tarefa.id,
                "titulo": tarefa.titulo,
                "descricao": tarefa.descricao,
                "estado": tarefa.estado,
                "data_criacao": tarefa.data_criacao,
                "data_limite": tarefa.data_limite
            }
            tarefas.append(tarefa_dict)

        return tarefas

    except Exception as e:
        print(f"Erro ao listar tarefas via SOAP: {e}")
        return []

def criar_tarefa(titulo, descricao, estado, data_criacao, data_limite):
    try:
        resposta = client.service.criar_tarefa(titulo, descricao, estado, data_criacao, data_limite)
        print("Resposta:", resposta)
        return resposta
    except Exception as e:
        print(f"Erro ao criar tarefa via SOAP: {e}")
        return None

# (Opcional) exemplo de envio manual de XML SOAP, se necessário
def processar_requisicao(xml_data):
    import requests
    headers = {'Content-Type': 'text/xml'}  # SOAP 1.1 usa 'text/xml'
    response = requests.post("http://localhost:8002", data=xml_data, headers=headers)
    return response.text

File: client/templates/soapClient.js
const SOAP_URL = 'http://127.0.0.1:8000/?wsdl';

// Função para listar tarefas
function listarTarefas() {
    const requestXml = `
        <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:web="http://localhost:8000/">
            <soapenv:Header/>
            <soapenv:Body>
                <web:listar_tarefas/>
            </soapenv:Body>
        </soapenv:Envelope>
    `;
    
    fetch(SOAP_URL, {
        method: "POST",
        headers: {
            "Content-Type": "text/xml;charset=UTF-8"
        },
        body: requestXml
    })
    .then(response => response.text())
    .then(xml => {
        const parser = new DOMParser();
        const xmlDoc = parser.parseFromString(xml, "text/xml");
        const tarefas = xmlDoc.getElementsByTagName("tarefa");
        
        const taskList = document.getElementById('taskList');
        taskList.innerHTML = "";
        
        Array.from(tarefas).forEach(tarefa => {
            taskList.innerHTML += `
                <p>
                    ID: ${tarefa.getElementsByTagName("id")[0].textContent},
                    Título: ${tarefa.getElementsByTagName("titulo")[0].textContent},
                    Estado: ${tarefa.getElementsByTagName("estado")[0].textContent},
                    Data Limite: ${tarefa.getElementsByTagName("data_limite")[0].textContent}
                </p>
            `;
        });
    })
    .catch(error => console.error("Erro ao listar tarefas:", error));
}

// Função para criar uma nova tarefa
function criarTarefa() {
    const titulo = document.getElementById('titulo').value;
    const descricao = document.getElementById('descricao').value;
    const estado = document.getElementById('estado').value;
    const dataCriacao = document.getElementById('data_criacao').value;
    const dataLimite = document.getElementById('data_limite').value;

    const requestXml = `
        <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:web="http://localhost:8000/">
            <soapenv:Header/>
            <soapenv:Body>
                <web:criar_tarefa>
                    <titulo>${titulo}</titulo>
                    <descricao>${descricao}</descricao>
                    <estado>${estado}</estado>
                    <data_criacao>${dataCriacao}</data_criacao>
                    <data_limite>${dataLimite}</data_limite>
                </web:criar_tarefa>
            </soapenv:Body>
        </soapenv:Envelope>
    `;

    fetch(SOAP_URL, {
        method: "POST",
        headers: {
            "Content-Type": "text/xml;charset=UTF-8"
        },
        body: requestXml
    })
    .then(response => response.text())
    .then(xml => {
        const parser = new DOMParser();
        const xmlDoc = parser.parseFromString(xml, "text/xml");
        const resposta = xmlDoc.getElementsByTagName("resposta")[0].textContent;
        alert(resposta);
    })
    .catch(error => console.error("Erro ao criar tarefa:", error));
}
